Mathlib/Tactic/NormNum/LegendreSymbol.lean

```diff
@@ -0,0 +1,473 @@
+
+theorem jacobiSymNat.qr₁' (a b : ℕ) (r : ℤ) (ha : a % 2 = 1) (hb : b % 4 = 1)
+    (hr : jacobiSymNat b a = r) : jacobiSymNat a b = r := by
+  rwa [jacobiSymNat, ← jacobiSym.quadratic_reciprocity_one_mod_four hb (Nat.odd_iff.mpr ha)]
+#align norm_num.jacobi_sym_nat.qr₁' Mathlib.Meta.NormNum.jacobiSymNat.qr₁'
+
+theorem jacobiSymNat.qr₁'_mod (a b ab : ℕ) (r : ℤ) (ha : a % 2 = 1) (hb : b % 4 = 1)
+    (hab : b % a = ab) (hr : jacobiSymNat ab a = r) : jacobiSymNat a b = r :=
+  jacobiSymNat.qr₁' _ _ _ ha hb <| jacobiSymNat.mod_left _ _ ab r hab hr
+#align norm_num.jacobi_sym_nat.qr₁'_mod Mathlib.Meta.NormNum.jacobiSymNat.qr₁'_mod
+
+theorem jacobiSymNat.qr₃ (a b : ℕ) (r : ℤ) (ha : a % 4 = 3) (hb : b % 4 = 3)
+    (hr : jacobiSymNat b a = r) : jacobiSymNat a b = -r := by
+  rwa [jacobiSymNat, jacobiSym.quadratic_reciprocity_three_mod_four ha hb, neg_inj]
+#align norm_num.jacobi_sym_nat.qr₃ Mathlib.Meta.NormNum.jacobiSymNat.qr₃
+
+theorem jacobiSymNat.qr₃_mod (a b ab : ℕ) (r : ℤ) (ha : a % 4 = 3) (hb : b % 4 = 3)
+    (hab : b % a = ab) (hr : jacobiSymNat ab a = r) : jacobiSymNat a b = -r :=
+  jacobiSymNat.qr₃ _ _ _ ha hb <| jacobiSymNat.mod_left _ _ ab r hab hr
+#align norm_num.jacobi_sym_nat.qr₃_mod Mathlib.Meta.NormNum.jacobiSymNat.qr₃_mod
+
+theorem isInt_jacobiSym : {a na : ℤ} → {b nb : ℕ} → {r : ℤ} →
+    IsInt a na → IsNat b nb → jacobiSym na nb = r → IsInt (jacobiSym a b) r
+  | _, _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, rfl => ⟨rfl⟩
+
+theorem isInt_jacobiSymNat : {a na : ℕ} → {b nb : ℕ} → {r : ℤ} →
+    IsNat a na → IsNat b nb → jacobiSymNat na nb = r → IsInt (jacobiSymNat a b) r
+  | _, _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, rfl => ⟨rfl⟩
+
+
+end Mathlib.Meta.NormNum
+
+end Lemmas
+
+section Evaluation
+
+/-!
+### Certified evaluation of the Jacobi symbol
+
+The following functions recursively evaluate a Jacobi symbol and construct the
+corresponding proof term.
+-/
+
+
+namespace Mathlib.Meta.NormNum
+
+open Lean Elab Tactic Qq
+
+/-- This evaluates `r := jacobiSymNat a b` recursively using quadratic reciprocity
+and produces a proof term for the equality, assuming that `a < b` and `b` is odd. -/
+partial def proveJacobiSymOdd (ea eb : Q(ℕ)) : (er : Q(ℤ)) × Q(jacobiSymNat $ea $eb = $er) :=
+  match eb.natLit! with
+  | 1 =>
+    show (er : Q(ℤ)) × Q(jacobiSymNat $ea 1 = $er) from
+      ⟨mkRawIntLit 1, q(jacobiSymNat.one_right $ea)⟩
+  | b =>
+    match ea.natLit! with
+    | 0 =>
+      have hb : Q(Nat.beq ($eb / 2) 0 = false) := (q(Eq.refl false) : Expr)
+      show (er : Q(ℤ)) × Q(jacobiSymNat 0 $eb = $er) from
+        ⟨mkRawIntLit 0, q(jacobiSymNat.zero_left $eb $hb)⟩
+    | 1 =>
+      show (er : Q(ℤ)) × Q(jacobiSymNat 1 $eb = $er) from
+        ⟨mkRawIntLit 1, q(jacobiSymNat.one_left $eb)⟩
+    | a =>
+      match a % 2 with
+      | 0 =>
+        match a % 4 with
+        | 0 =>
+          have ha : Q(Nat.mod $ea 4 = 0) := (q(Eq.refl 0) : Expr)
+          have hb : Q(Nat.mod $eb 2 = 1) := (q(Eq.refl 1) : Expr)
+          have c := a / 4
```

Comment:
You could inline quite a few of these haves into the (sole) place the definition is used. If you think it hurts readability I guess we can leave it.